### PR TITLE
Safety and compliance

### PR DESCRIPTION
**Pull request by KyleMacPherson**
https://trello.com/c/uo3Po1he/970-5-edit-safety-and-compliance-information

## Description
Allow users to edit safety and compliance information for a case

## Checklist:
- [ ] Have you documented your changes in the pull request description?
- [ ] Does the change present any security considerations?
- [ ] Is any gem functionality overloaded? Eg: Devise controller methods being overloaded.
- [ ] Has acceptance criteria been tested by a peer?
- [ ] Has the CHANGELOG been updated? (If change is worth telling users about.)

### General testing (author)
- [ ] Test without JavaScript
- [ ] Test on small screen

### Accessibility testing (author)
- [ ] Works keyboard only
- [ ] Tested with one screen reader
- [ ] Zoom page to 400% - content still visible
- [ ] Disable CSS - does content make sense and appear in a logical order?

### CODE REVIEW

**Comment by nasirkhanbeis** [comment]
spotted few issues
1. safe and compliant is default selected when I click change for existing OPSS cases
2. When I don't change but simply click change, then I am not suppose to see confirmation banner

**Review by nasirkhanbeis — APPROVED** [review]
re-tested on review app 1266, looking good as expected.

File: app/services/change_safety_and_compliance_data.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      investigation.assign_attributes(hazard_description: hazard_description, hazard_type: hazard_type, non_compliant_reason: non_compliant_reason, reported_reason: reported_reason)
+    end
+
+    if reported_reason == :unsafe
+      investigation.assign_attributes(hazard_description: hazard_description, hazard_type: hazard_type, non_compliant_reason: nil, reported_reason: reported_reason)
+    end
+
+    if reported_reason == :non_compliant
+      investigation.assign_attributes(hazard_description: nil, hazard_type: nil, non_compliant_reason: non_compliant_reason, reported_reason: reported_reason)
+    end
```
Comment (StupidCodeFactory) [inline app/services/change_safety_and_compliance_data.rb:43]: Suggestion: all the logic `should/could` be calculated in the form object. Seems like it form logic creeping up in the service class. but it's debatable.

File: app/services/change_safety_and_compliance_data.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    AuditActivity::Investigation::ChangeSafetyAndComplianceData
+  end
+
+  def user_source
+    @user_source ||= UserSource.new(user: user)
+  end
+
+  def send_notification_email(investigation, user)
+    email_recipients_for_team_with_access(investigation, user).each do |entity|
+      email = entity.is_a?(Team) ? entity.team_recipient_email : entity.email
```
Comment (StupidCodeFactory) [inline app/services/change_safety_and_compliance_data.rb:66]: ✂️

File: app/services/change_safety_and_compliance_data.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      NotifyMailer.investigation_updated(
+        investigation.pretty_id,
+        entity.name,
+        email,
```
Comment (StupidCodeFactory) [inline app/services/change_safety_and_compliance_data.rb:70]:
```suggestion
        entity.email,
```

`Team#email` is aliased to team_recipeint_email

File: spec/features/change_safety_and_compliance_spec.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
+require "rails_helper"
+
+RSpec.feature "Change safety and compliance details for a case", :with_stubbed_mailer, :with_stubbed_elasticsearch do
+  let(:user)           { create(:user, :activated, team: create(:team, name: "Portsmouth Trading Standards"), name: "Bob Jones") }
+  let(:investigation)  { create(:allegation, creator: user, reported_reason: "unsafe_and_non_compliant", hazard_type: "Burns", hazard_description: "FIRE FIRE FIRE", non_compliant_reason: "Covered in petrol") }
```
Comment (StupidCodeFactory) [inline spec/features/change_safety_and_compliance_spec.rb:5]:
```suggestion
  let(:investigation)  { create(:allegation, creator: user, reported_reason: "unsafe_and_non_compliant", hazard_type: "Burns", hazard_description: "FIRE! FIRE! FIRE! ", non_compliant_reason: "Covered in petrol") }
```

File: spec/services/change_safety_and_compliance_data_spec.rb [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+          it "updates hazard_description" do
+            expect(investigation.hazard_description).to eq(hazard_description)
+          end
+
+          it "makes non_compliant_reason nil" do
+            expect(investigation.non_compliant_reason).to eq(nil)
+          end
+
+          it "creates an activity entry" do
+            expect(activity_entry.metadata).to eql({
```
Comment (StupidCodeFactory) [inline spec/services/change_safety_and_compliance_data_spec.rb:177]:
```suggestion
            expect(activity_entry.metadata).to eq({
```

**Review by StupidCodeFactory — CHANGES_REQUESTED** [review]

File: app/controllers/investigations/safety_and_compliance_controller.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+      if @why_reporting_form.valid?
+        result = ChangeSafetyAndComplianceData.call!(
+          @why_reporting_form.serializable_hash.merge({
+            investigation: investigation,
+            user: current_user
+          })
+        )
+
+        flash[:success] = "Case information changed." if result.changes_made
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/controllers/investigations/safety_and_compliance_controller.rb:23]: We do this elsewhere, for example the case status (open/close case) which I'm working on currently.

File: spec/features/change_safety_and_compliance_spec.rb (hunk 4 above)
Comment (StupidCodeFactory) [reply spec/features/change_safety_and_compliance_spec.rb:5]: https://www.youtube.com/watch?v=1EBfxjSFAxQ

**Review by StupidCodeFactory — APPROVED** [review]

**Review by tobyhughes — APPROVED** [review]
Checked:
- New section on overview tab
- Correct variables showing (user input for TS cases and `Not provided` for OPSS cases where the input is not required as part of case creation)
- Edit permissions (i.e. can only edit on cases where added to the team with edit permissions)
- Activity timeline
- Case export

All working as expected

**Pull request merged**